### PR TITLE
[JENKINS-28071] generalise CLI's BuildCommand target from AbstractProject to Job

### PR DESCRIPTION
**Pull request by corngood**
[JENKINS-28071](https://issues.jenkins-ci.org/browse/JENKINS-28071)

This allows e.g. WorkflowJob to build, fixing symptom [JENKINS-29826](https://issues.jenkins-ci.org/browse/JENKINS-29826)

`JobOptionHandler` is based on `AbstractProjectOptionHandler`.  I'm not sure if the latter is still needed. The absence of `JobOptionHandler` only caused a runtime error, so I'm reluctant to remove it.

I didn't yet put any effort into abstracting (or finding an abstraction for) `isDisabled` or `poll`.

TODO: tests

### CODE REVIEW

File: core/src/main/java/hudson/cli/handlers/JobOptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2004-2009, Sun Microsystems, Inc.
```
Comment (KostyaSha) [inline core/src/main/java/hudson/cli/handlers/JobOptionHandler.java]: ^

Comment (corngood) [inline core/src/main/java/hudson/cli/handlers/JobOptionHandler.java]: This is based on a file with this license and (c).  What do you expect here?

Comment (jtnord) [inline core/src/main/java/hudson/cli/handlers/JobOptionHandler.java]: as this is a new file the date should be the current year - and the copyright owner should be who owns the copyright of this code (either you or the company you work for depending on when you did this code change and your employment contract)

e..g 
`Copyright (c) 2016, David McFarland` or `Copyright (c) 2016, Acme Inc.` (if you work for Acme Inc,)

Comment (corngood) [inline core/src/main/java/hudson/cli/handlers/JobOptionHandler.java]: "The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software."

Since this file is essentially a rename I thought it was appropriate to leave it intact (the dates were never changed in other revisions).

If I had written it from scratch I never would have put a copyright/license on it at all (as in TopLevelItemOptionHandler.java).  I can take it off completely if that's better.

Comment (jtnord) [inline core/src/main/java/hudson/cli/handlers/JobOptionHandler.java]: Ok I missed this was a a rename.
Probably should have the date range updated and you added as a contributor, but I think if this was just a rename then we are ok

Comment (corngood) [inline core/src/main/java/hudson/cli/handlers/JobOptionHandler.java]: I didn't bother doing that since nobody else has when they worked on the source file...  Doesn't matter to me though.

**Comment by corngood** [comment]
We've been running this for a while with no problems, and I haven't had any comments on the implementation, so removing WIP.  Do I need to do anything to remove the tag?

**Comment by corngood** [comment]
@oleg-nenashev Should I reopen to remove the WIP flag?  This is ready to merge IMO.

**Comment by corngood** [comment]
@jglick Could I get your input on this, since it's meant to allow CLI builds of workflows?

File: core/src/main/java/hudson/cli/BuildCommand.java [hunk 2]
```diff
@@ -134,15 +140,15 @@ protected int run() throws Exception {
             a = new ParametersAction(values);
         }
 
-        if (checkSCM) {
-            if (job.poll(new StreamTaskListener(stdout, getClientCharset())).change == Change.NONE) {
+        if (checkSCM && job instanceof AbstractProject<?, ?>) {
+            if (((AbstractProject<?, ?>)job).poll(new StreamTaskListener(stdout, getClientCharset())).change == Change.NONE) {
```
Comment (amuniz) [inline core/src/main/java/hudson/cli/BuildCommand.java]: You can run polling in general by using `SCMTriggerItems.asSCMTriggerItem(job).poll(listener).hasChanges()`

**Comment by amuniz** [comment]
LGTM :+1:

**Comment by amuniz** [comment]
Is something preventing to merge this?
@oleg-nenashev do you think it's still in-progress?

**Comment by jglick** [comment]
First I have heard of it, I can take a look.

File: core/src/main/java/hudson/cli/BuildCommand.java (hunk 2 above)
Comment (jglick) [inline core/src/main/java/hudson/cli/BuildCommand.java]: Right, this needs to be fixed.

File: core/src/main/java/hudson/cli/BuildCommand.java [hunk 3]
```diff
@@ -134,15 +140,15 @@ protected int run() throws Exception {
...
                 return 0;
             }
         }
 
         if (!job.isBuildable()) {
             String msg = Messages.BuildCommand_CLICause_CannotBuildUnknownReasons(job.getFullDisplayName());
-            if (job.isDisabled()) {
+            if (job instanceof AbstractProject<?, ?> && ((AbstractProject<?, ?>)job).isDisabled()) {
```
Comment (jglick) [inline core/src/main/java/hudson/cli/BuildCommand.java]: JENKINS-27299 for reference, but anyway this only comes down to the choice of a message.

File: core/src/main/java/hudson/cli/BuildCommand.java (hunk 2 above)
Comment (jglick) [inline core/src/main/java/hudson/cli/BuildCommand.java]: For example

```java
if (checkSCM) {
    SCMTriggerItem pollable = SCMTriggerItems.asSCMTriggerItem(job);
    if (pollable != null && pollable.poll(new StreamTaskListener(stdout, getClientCharset())).change == Change.NONE) {
```

(I suspect the original code ought to have been using `hasChanges()` to deal with `INSIGNIFICANT`, but anyway that is outside the scope of this PR.)

**Comment by jglick** [comment]
> `JobOptionHandler` is based on `AbstractProjectOptionHandler`. I'm not sure if the latter is still needed.

I think so, in case other commands continue to request an `AbstractProject` specifically.

**Comment by jglick** [comment]
Still in progress until the `poll` call is fixed, I suppose.

**Comment by corngood** [comment]
Updated with poll suggestion.  This compiles, but I haven't tested it myself.

**Comment by amuniz** [comment]
Tests are passing, so LGTM - again - :+1:

**Comment by corngood** [comment]
Bump.  Who do I need to merge this?

**Comment by corngood** [comment]
@jglick  Can you help me out with this?  Seems to be in limbo.

**Comment by oleg-nenashev** [comment]
Fixed the label. Just for the future: please do not squash before the review finishes. It does not allow reviewing diffs after comments

File: core/src/main/java/hudson/cli/BuildCommand.java (hunk 3 above)
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/BuildCommand.java]: Maybe it's a subject for the core enhancement. I see no reason why other Job types have no such method in API

**Comment by oleg-nenashev** [comment]
👍 from me

File: core/src/main/java/hudson/cli/BuildCommand.java [hunk 4]
```diff
@@ -135,14 +142,14 @@ protected int run() throws Exception {
         }
 
         if (checkSCM) {
-            if (job.poll(new StreamTaskListener(stdout, getClientCharset())).change == Change.NONE) {
+            if (!SCMTriggerItem.SCMTriggerItems.asSCMTriggerItem(job).poll(new StreamTaskListener(stdout, getClientCharset())).hasChanges()) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/BuildCommand.java]: asSCMTriggerItem(job) may return null, needs check

**Comment by oleg-nenashev** [comment]
Reverted my +1. There is a potential NPE

**Comment by corngood** [comment]
@oleg-nenashev  Like so?

Let me know if you want me to squash it.

File: core/src/main/java/hudson/cli/BuildCommand.java [hunk 5]
```diff
@@ -135,14 +142,15 @@ protected int run() throws Exception {
         }
 
         if (checkSCM) {
-            if (job.poll(new StreamTaskListener(stdout, getClientCharset())).change == Change.NONE) {
+            SCMTriggerItem item = SCMTriggerItem.SCMTriggerItems.asSCMTriggerItem(job);
+            if (item != null && !item.poll(new StreamTaskListener(stdout, getClientCharset())).hasChanges()) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/BuildCommand.java]: If it is null, we cannot check SCM. Maybe makes sense to fail the command in such case

**Comment by corngood** [comment]
@oleg-nenashev Now with exception for missing SCM trigger

**Comment by oleg-nenashev** [comment]
👍

**Comment by corngood** [comment]
@oleg-nenashev Should I squash it?  Is there anything else I can do to help get this merged?

Cheers

**Comment by oleg-nenashev** [comment]
@corngood Everything is fine, it's possible to squash via GitHub web UI now. I'm just waiting for the feedback from somebody, we commonly merge after two+ votes or after a timeout. If there is no responses, I'll merge it tomorrow.

 CC @amuniz @jglick @jtnord

**Comment by oleg-nenashev** [comment]
@corngood 
Thanks for your patience! It's integrated now
